ts/components/search/SearchButton.tsx

```diff
@@ -82,12 +83,19 @@ class SearchButton extends React.Component<Props, State> {
   }
 
   private handleSearchPress = () => {
-    const { searchText } = this.state;
-    this.setState({
-      searchText: some("")
-    });
-    this.props.dispatchSearchText(searchText);
-    this.props.dispatchSearchEnabled(true);
+    const { onSearch } = this.props;
+
+    fromNullable(onSearch).foldL(
```

Comment:
What is the reason of introducing `onSearch`?
I guess the old one use the store. The current one `onSearch` is a callback without the searched text. What is the difference?

In this PR it seems to be not used

Reply:
@Undermaken The name is not self-explaining, I understand your thought.

onSearch is used by `CgnMerchantsTabScreen` [here](https://github.com/pagopa/io-app/pull/3033/files#diff-661ab4f9284847acb617c19511723f44a281d6f4f3a5319d0d0f321e99a38a34R89)

The function handles a custom search component from the Header of the app, since the filter modal can be opened both from the footer CTA and header search button/icon

Reply:
Thx. Now it's all clear.
So my proposal is to add this new callback `onSearch` without altering the current behavior 
Ad a developer if I use this component by passing `onSearch` these actions won't be dispatched

```typescript
this.props.dispatchSearchText(searchText);
this.props.dispatchSearchEnabled(true);
```

Is it right?
If yes what do you think to do both (dispatch and onSearch)?
```typescript
private handleSearchPress = () => {
    const { onSearch } = this.props;
    const { searchText } = this.state;
    this.setState({
      searchText: some("")
    });
    this.props.dispatchSearchText(searchText);
    this.props.dispatchSearchEnabled(true);
    onSearch?.();
  };
```

Reply:
I have been thinking about this proposal, but what are the advantages to dispatch those 2 actions when we're handling another custom behaviour from the same button?

For example this is the code from `dispatchSearchEnabled`:
```
(isSearchEnabled: boolean) => {
    const searchType = props.searchType;
    switch (searchType) {
      case "Messages":
        dispatch(searchMessagesEnabled(isSearchEnabled));
        break;
      case "Services":
        dispatch(searchServicesEnabled(isSearchEnabled));
        break;
    }
  }
```

Dispatching it we would save an incorrect information on redux store.

Instead setting state to 
```
this.setState({
  searchText: some("")
});
```
would affect the UI changing the button with an Input component

Reply:
Maybe a better solution would be renaming the function with `onCustomSearch` or something similar?

Reply:
I'm try to dimostrate that this code is altering the previous behavior:
now if a developer passes `onSearch` handler to that component, it doesn't dispatch the actions (`dispatchSearchText` + `dispatchSearchEnabled`) anymore

```typescript
fromNullable(onSearch).foldL(
      () => {
        const { searchText } = this.state;
        this.setState({
          searchText: some("")
        });
        this.props.dispatchSearchText(searchText);
        this.props.dispatchSearchEnabled(true);
      },
      os => os()
    );
```
So as a developer I must know that if I pass a particular prop to the component, it will change its behavior and no action will be dispatched.
This is the proof
![Schermata 2021-05-20 alle 12 39 14](https://user-images.githubusercontent.com/822471/118964971-68262600-b968-11eb-8d53-816ec06f62ce.png)
Added in `ts/components/screens/TopScreenComponent.tsx`
produces this

https://user-images.githubusercontent.com/822471/118965014-75dbab80-b968-11eb-9f46-09593166ddcc.mov

So my suggestion is to execute both `dispatch` + `onSearch`. `searchType` could be undefined so this code won't produce side effects
```typescript
const searchType = props.searchType;
    switch (searchType) {
      case "Messages":
        dispatch(searchMessagesEnabled(isSearchEnabled));
        break;
      case "Services":
        dispatch(searchServicesEnabled(isSearchEnabled));
        break;
    }
```

Furthermore `onSearch` could be `onSearchTap`. What do you think?

Reply:
^^ @CrisTofani

Reply:
@Undermaken I actually understand your point, but executing both would cause another side effect here reported
![Simulator Screen Recording - iPhone 11 - 2021-05-21 at 15 38 59](https://user-images.githubusercontent.com/3959405/119146404-e22edb80-ba4a-11eb-97f8-477fb07b5966.gif)

Apart from dispatching the actions it sets the local state of the component with:
```
        this.setState({
          searchText: some("")
        });
```
that removes the button adding an input text...

What's your proposal about this?